File: internal/probe/tcp.go
package probe

import (
	"context"
	"fmt"
	"net"
	"time"
)

// TCP probes a host by timing a TCP connect to a port.
// It needs no privileges and works through most firewalls.
type TCP struct {
	addr    string
	port    string
	timeout time.Duration
}

// NewTCP returns a TCP prober for an already-resolved address and port.
func NewTCP(ip *net.IPAddr, port string, timeout time.Duration) *TCP {
	return &TCP{addr: net.JoinHostPort(ip.IP.String(), port), port: port, timeout: timeout}
}

func (p *TCP) Kind() string { return "tcp :" + p.port }

func (p *TCP) Probe(ctx context.Context, seq int) Result {
	at := time.Now()
	d := net.Dialer{Timeout: p.timeout}
	start := time.Now()
	conn, err := d.DialContext(ctx, "tcp", p.addr)
	rtt := time.Since(start)
	if err != nil {
		return Result{Seq: seq, Err: fmt.Errorf("connect: %w", err), At: at}
	}
	_ = conn.Close()
	return Result{Seq: seq, RTT: rtt, At: at}
}

File: src/probe/tcp.rs
//! Probes a host by timing a TCP connect to a port.
//!
//! Needs no privileges and works through most firewalls.

use std::net::{IpAddr, SocketAddr, TcpStream};
use std::time::{Duration, Instant};

use crate::probe::{Prober, Sample};

pub struct Tcp {
    addr: SocketAddr,
    port: u16,
    timeout: Duration,
}

impl Tcp {
    /// Returns a TCP prober for an already-resolved address and port.
    pub fn new(ip: IpAddr, port: u16, timeout: Duration) -> Self {
        Self {
            addr: SocketAddr::new(ip, port),
            port,
            timeout,
        }
    }
}

impl Prober for Tcp {
    fn kind(&self) -> String {
        format!("tcp :{}", self.port)
    }

    fn probe(&self, seq: usize) -> Sample {
        let start = Instant::now();
        match TcpStream::connect_timeout(&self.addr, self.timeout) {
            Ok(_) => Sample::reply(seq, start.elapsed()),
            Err(_) => Sample::lost(seq),
        }
    }
}
